File: app/src/main/java/com/boost/espressotest/data/api/tools/ConnectivityInterceptor.java
package com.boost.espressotest.data.api.tools;

import android.content.Context;
import android.support.annotation.NonNull;

import com.boost.espressotest.presentation.tools.Utils;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * @author PerSpiKyliaTor on 11.04.2017.
 */

public class ConnectivityInterceptor implements Interceptor {

    private Context mContext;

    // TODO: 1/24/18 create NetworkUtils(context) class with method isOnline and provide it here to avoid static methods (for testing purposes)
    public ConnectivityInterceptor(Context context) {
        mContext = context;
    }

    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        if (!Utils.isOnline(mContext)) {
            throw new NoConnectivityException();
        }

        Request.Builder builder = chain.request().newBuilder();
        return chain.proceed(builder.build());
    }
}

File: app/src/main/java/com/boost/espressotest/data/api/tools/ConnectivityInterceptor.kt
package com.boost.espressotest.data.api.tools

import android.content.Context
import com.boost.espressotest.domain.exceptions.NoConnectivityException
import com.boost.espressotest.presentation.tools.isOnline
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

class ConnectivityInterceptor(private val context: Context) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        if (!context.isOnline()) {
            throw NoConnectivityException()
        }

        val builder = chain.request().newBuilder()
        return chain.proceed(builder.build())
    }
}